src/c4/yml/emit.hpp

```diff
@@ -241,6 +240,221 @@ CharOwningContainer emitrs(NodeRef const& n)
+{
+public:
+
+    using Writer::Writer;
+
+    /** emit JSON.
+     *
+     * When writing to a buffer, returns a substr of the emitted JSON.
+     * If the given buffer has insufficient space, the returned span will
+     * be null and its size will be the needed space. No writes are done
+     * after the end of the buffer.
+     *
+     * When writing to a file, the returned substr will be null, but its
+     * length will be set to the number of bytes written. */
+    substr emit(Tree const& t, size_t id, bool error_on_excess);
+    /** @overload */
+    substr emit(Tree const& t, bool error_on_excess=true) { return emit(t, t.root_id(), error_on_excess); }
+    /** @overload */
+    substr emit(NodeRef const& n, bool error_on_excess=true) { return emit(*n.tree(), n.id(), error_on_excess); }
+
+    size_t tell() const;
+    void   seek(size_t p);
+
+private:
+
+    size_t  _visit(Tree const& t, size_t id);
+    void _do_visit(Tree const& t, size_t id);
+
+private:
+
+    template<class T>
+    C4_ALWAYS_INLINE void _write(T a)
+    {
+        this->Writer::_do_write(a);
+    }
+
+    template<size_t N>
+    C4_ALWAYS_INLINE void _write(const char (&a)[N])
+    {
+        csubstr s(a);
+        this->Writer::_do_write(s);
+    }
+
+    void _write(NodeScalar const& sc, NodeType flags);
+
+    void _write_scalar(csubstr s);
+
+    enum {
+        _keysc =  (KEY)  | ~(VAL),
+        _valsc = ~(KEY)  |  (VAL),
+    };
+
+    C4_ALWAYS_INLINE void _writek(Tree const& t, size_t id) { _write(t.keysc(id), t.type(id) & ~(VAL)); }
+    C4_ALWAYS_INLINE void _writev(Tree const& t, size_t id) { _write(t.valsc(id), t.type(id) & ~(KEY)); }
+};
+
+//-----------------------------------------------------------------------------
+//-----------------------------------------------------------------------------
+//-----------------------------------------------------------------------------
+
+/** emit JSON to the given file. A null file defaults to stdout.
+ * Return the number of bytes written. */
+inline size_t emit_json(Tree const& t, size_t id, FILE *f)
+{
+    EmitterJSONFile em(f);
+    size_t len = em.emit(t, id, /*error_on_excess*/true).len;
+    return len;
+}
+
+/** emit JSON to the given file. A null file defaults to stdout.
+ * Return the number of bytes written.
+ * @overload */
+inline size_t emit_json(Tree const& t, FILE *f=nullptr)
+{
+    return emit_json(t, t.root_id(), f);
+}
+
+/** emit JSON to the given file. A null file defaults to stdout.
+ * Return the number of bytes written.
+ * @overload */
+inline size_t emit_json(NodeRef const& r, FILE *f=nullptr)
+{
+    return emit_json(*r.tree(), r.id(), f);
+}
+
+
+
+struct as_json
+{
+    NodeRef node;
+    as_json(NodeRef const& n) : node(n) {}
+    as_json(Tree const& t) : node(t.rootref()) {}
+};
+template<class OStream>
+inline OStream& operator<< (OStream& s, as_json const& js)
+{
+    EmitterJSONOStream<OStream> em(*s.tree); // instantiate a json emitter instead
```

Comment:
this can't compile. `s` is a stream, and has no tree member. The tests should cover this.

Reply:
fixed, test added